libgalois/include/galois/runtime/Context.h

```diff
@@ -44,10 +42,35 @@ enum ConflictFlag {
   BREAK            = 2
 };
 
-#ifdef GALOIS_USE_LONGJMP_ABORT
-extern thread_local jmp_buf execFrame;
+extern thread_local std::jmp_buf execFrame;
+
+class Lockable;
+
+[[noreturn]] inline void signalConflict(Lockable* = nullptr) {
```

Comment:
The reason for moving these to the header is that it means that users of libgalois can use either kind of aborts based only on the preprocessor flags used in their build and not on the way the original libgalois was built. If we include these functions inside libgalois we loose that.

Reply:
Hmm... sorry, I was context switching too much :D This seems fine.

FWIW, what I meant was:

```
[[noreturn]] inline void signalConflict(...) {
#if ...
  signalConflictException(...)
#else 
  signalConflictAbort(...)
#endif
}
```
as that mirrors the original code's structure but I have no reason to believe that there is a performance difference with what is in the PR currently. Here are the icache differences: https://godbolt.org/z/ndskU6 They mainly affect the throw case.


So does https://github.com/IntelligentSoftwareSystems/Galois/pull/68/files/74d1bd74e78ec292c7d89edbf65ab5d057f1366b#diff-89fc89ea51a7b13fe026cbd1781735abR156 need to be in the header as well?

Reply:
Yep, that needs to be in the header too because it calls these functions. Otherwise an abort of the type that's set when Galois is built happens inside libgalois and that could cause a mismatch with the way the user has configured the headers.

Reply:
Oh, that makes sense. Yes we could provide two different variants of these functions in libgalois and then just pick which one to use in the headers. I don't have a strong preference there.

Reply:
I remember back in the day we spent a lot of time optimizing the code layout for the successful acquire case because that is on the critical path. This is why `acquire` is not inlined in the first place. How `signalConflict` is implemented matters much less because it is only called from `acquire` itself.

There is a secondary maintainability issue which is how to avoid baking in the abort implementation method since any cpp file that calls signalConflict or any of its callers will break the feature we are trying to build. I was hoping that splitting the names would help (`fooException()`, `fooJump`) but I don't think it does. Given that in the long run we need to ween ourselves of longjmp (because it breaks RAII), if there isn't a nice and easy fix for this maintainability issue, either approach seems fine to me.